### ISSUE TITLE
ratchet 0.2.3 fails in a GitHub Codespace

### ISSUE THREAD

**Issue by artis3n** [sethvargo/ratchet#33]
### TL;DR

Running `ratchet pin .github/workflows/<GITHUB ACTIONS WORKFLOW FILE>.yml` fails under a GitHub Codespace due to file ownership permission issues (`chown` error), although I can't really tell why. Both the executable and the file + directory are owned by the `vscode` user.

### Expected behavior

`ratchet pin` would successfully execute inside the permissions environment of a Codespace.

### Observed behavior

Ran the following to put ratchet onto the Codespace:
```sh
wget -O /tmp/ratchet.tar.gz https://github.com/sethvargo/ratchet/releases/download/v0.2.3/ratchet_0.2.3_linux_amd64.tar.gz && \
tar -xzf /tmp/ratchet.tar.gz -C /tmp && \
sudo mv /tmp/ratchet /usr/local/bin/ratchet
```
![image](https://user-images.githubusercontent.com/6969296/178363131-54deb5e9-0052-4468-92a5-ce8f9a949e0b.png)

Ratchet binary:
![image](https://user-images.githubusercontent.com/6969296/178363198-74921a02-6947-485f-bb88-cc507bcb7913.png)

File permissions of the `.github/workflows` file and directory where I'm then running ratchet:
![image](https://user-images.githubusercontent.com/6969296/178363286-6a071832-0035-4034-bdca-bf56ee2a517b.png)

Running ratchet results in:
```
failed to save .github/workflows/ci.yml: failed to save file .github/workflows/ci.yml: failed to preserve file permissions: failed to change ownership: chown .github/workflows/ratchet-3471756679: operation not permitted
```
![image](https://user-images.githubusercontent.com/6969296/178363337-752fe5fe-e916-4ce0-be7e-babe5d76f605.png)

For sanity's sake, also with an absolute path:
![image](https://user-images.githubusercontent.com/6969296/178363449-4d939dcf-bafa-41c5-860e-7e2fe60bc3bc.png)

Running `sudo chown vscode:root /usr/local/bin/ratchet` to exactly match the directory and file permissions doesn't work either and results in the same error, but shouldn't matter.

### Version

ratchet v0.2.3 (7f1ec2c2a3482c2e061b9dfd6c36de705f0ad81c)

### On what operating system are you using ratchet?

Linux

### CI/CD workflow file

```yaml
---
name: CI Tests

on:
  pull_request_target:
    branches:
     - main

env:
  # https://www.jeffgeerling.com/blog/2020/getting-colorized-output-molecule-and-ansible-on-github-actions-ci
  PY_COLORS: '1'
  ANSIBLE_FORCE_COLOR: '1'

jobs:
  molecule:
    environment: E2E
    runs-on: ubuntu-latest

    strategy:
      matrix:
        distros:
          - geerlingguy/docker-centos7-ansible:latest
          - geerlingguy/docker-rockylinux8-ansible:latest
          - ghcr.io/artis3n/docker-almalinux8-ansible:latest
          - ghcr.io/artis3n/docker-rhel8-ansible:latest
          - ghcr.io/artis3n/docker-oraclelinux8-ansible:latest
          - ghcr.io/artis3n/docker-amazonlinux2-ansible:latest
          - ghcr.io/artis3n/docker-ubuntu2204-ansible:latest
          - geerlingguy/docker-ubuntu2004-ansible:latest
          - geerlingguy/docker-ubuntu1804-ansible:latest
          - geerlingguy/docker-ubuntu1604-ansible:latest
          - ghcr.io/artis3n/docker-debian12-ansible:latest
          - geerlingguy/docker-debian11-ansible:latest
          - geerlingguy/docker-debian10-ansible:latest
          - geerlingguy/docker-debian9-ansible:latest
          - geerlingguy/docker-fedora31-ansible:latest
          - ghcr.io/artis3n/docker-arch-ansible:latest
      fail-fast: false

    steps:
      - uses: actions/checkout@v3
        with:
          ref: ${{ github.event.pull_request.head.sha }}

      # - name: Install dependency manager
      #   run: pipx install poetry

      - name: Set up Python 3.x
        id: setup-python
        uses: actions/setup-python@v4.0.0
        with:
          python-version: "3.10"
          # Once supported in https://github.com/actions/setup-python/pull/281
          # cache: 'poetry'

      - name: Install dependency manager
        run: pip install poetry

      - name: Install packages
        run: poetry install --no-interaction

      - name: Lint
        run: make lint

      - name: Default scenario
        run: poetry run molecule test --scenario-name default
        env:
          MOLECULE_DISTRO: "${{ matrix.distros }}"
          TAILSCALE_CI_KEY: "${{ secrets.TAILSCALE_CI_KEY }}"

      - name: Uninstall scenario
        run: poetry run molecule test --scenario-name state-absent
        env:
          MOLECULE_DISTRO: "${{ matrix.distros }}"
          TAILSCALE_CI_KEY: "${{ secrets.TAILSCALE_CI_KEY }}"

  molecule-skip-auth:
    environment: E2E
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v3
        with:
          ref: ${{ github.event.pull_request.head.sha }}

      # - name: Install dependency manager
      #   run: pipx install poetry

      - name: Set up Python 3.x
        id: setup-python
        uses: actions/setup-python@v4.0.0
        with:
          python-version: "3.10"
          # Once supported in https://github.com/actions/setup-python/pull/281
          # cache: 'poetry'

      - name: Install dependency manager
        run: pip install poetry

      - name: Install packages
        run: poetry install --no-interaction

      - name: Molecule - Skip Authentication
        run: poetry run molecule test --scenario-name skip-authentication

  molecule-vars-priority:
    environment: E2E
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v3
        with:
          ref: ${{ github.event.pull_request.head.sha }}

      # - name: Install dependency manager
      #   run: pipx install poetry

      - name: Set up Python 3.x
        id: setup-python
        uses: actions/setup-python@v4.0.0
        with:
          python-version: "3.10"
          # Once supported in https://github.com/actions/setup-python/pull/281
          # cache: 'poetry'

      - name: Install dependency manager
        run: pip install poetry

      - name: Install packages
        run: poetry install --no-interaction

      - name: Molecule - Vars Priority Validation
        run: poetry run molecule test --scenario-name vars-priority
        env:
          TAILSCALE_CI_KEY: "${{ secrets.TAILSCALE_CI_KEY }}"

  molecule-args:
    environment: E2E
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v3
        with:
          ref: ${{ github.event.pull_request.head.sha }}

      # - name: Install dependency manager
      #   run: pipx install poetry

      - name: Set up Python 3.x
        id: setup-python
        uses: actions/setup-python@v4.0.0
        with:
          python-version: "3.10"
          # Once supported in https://github.com/actions/setup-python/pull/281
          # cache: 'poetry'

      - name: Install dependency manager
        run: pip install poetry

      - name: Install packages
        run: poetry install --no-interaction

      - name: Molecule - Vars Priority Validation
        run: poetry run molecule test --scenario-name args
        env:
          TAILSCALE_CI_KEY: "${{ secrets.TAILSCALE_CI_KEY }}"
```

### Relevant log output

```text
ratchet pin .github/workflows/ci.yml

failed to save .github/workflows/ci.yml: failed to save file .github/workflows/ci.yml: failed to preserve file permissions: failed to change ownership: chown .github/workflows/ratchet-3471756679: operation not permitted
```

### Additional information

Was playing around with setting up ratchet in this repo, so feel to open up a Codespace on a fork to exactly reproduce the environment I encountered this in.

https://github.com/artis3n/ansible-role-tailscale/tree/ratchet

**Comment by artis3n** [comment]
I didn't see it under the help flag, but is there a debug flag for ratchet to get better insight into what it's trying to do during this file saving that's resulting in this failure?

**Comment by sethvargo** [comment]
The issue appears to be the _group_. The group on your file is set to `root`, and ratchet doesn't have permission to change the group ownership to root.

**Comment by artis3n** [comment]
> The issue appears to be the _group_. The group on your file is set to `root`, and ratchet doesn't have permission to change the group ownership to root.

I tried that! Sorry, hidden among my comments:
> Running `sudo chown vscode:root /usr/local/bin/ratchet` to exactly match the directory and file permissions doesn't work either and results in the same error

Here's documentation of the attempt, note the file permissions:
![image](https://user-images.githubusercontent.com/6969296/178372115-488a1c09-5285-41b2-8371-b76b7c6dc4c5.png)

**Comment by artis3n** [comment]
It's very weird, it's standard file ownership Go code and all the file perms in the Codespace appear "correct." Not really sure why the error is occurring.

**Comment by sethvargo** [comment]
Ratchet attempts to preserve the permissions of the source file. In your screenshot, it's vscode:root. It needs to be vscode:vscode.

**Comment by artis3n** [comment]
Yeah, that's fair. I'm kinda limited by how the Codespace initializes its environment. I can make a `chown` on the workspace root and we'll see if that triggers unintended other issues in the environment.

Regardless, yeah, I don't see action for `ratchet` to take :) Will close
